flink-core/src/main/java/org/apache/flink/util/ExceptionUtils.java

```diff
@@ -110,51 +112,76 @@ public static boolean isJvmFatalOrOutOfMemoryError(Throwable t) {
+	 *                                       <code>null</code> will disable handling this class of error.
+	 * @param jvmDirectOomNewErrorMessage The message being used for direct memory-related OutOfMemoryErrors. Passing
+	 *                                    <code>null</code> will disable handling this class of error.
+	 * @param jvmHeapSpaceOomNewErrorMessage The message being used for Heap space-related OutOfMemoryErrors. Passing
+	 *                                       <code>null</code> will disable handling this class of error.
 	 */
-	@Nullable
-	public static Throwable tryEnrichOutOfMemoryError(
-			@Nullable Throwable exception,
-			String jvmMetaspaceOomNewErrorMessage,
-			String jvmDirectOomNewErrorMessage) {
-		boolean isOom = exception instanceof OutOfMemoryError;
-		if (!isOom) {
-			return exception;
-		}
-
-		OutOfMemoryError oom = (OutOfMemoryError) exception;
-		if (isMetaspaceOutOfMemoryError(oom)) {
-			return changeOutOfMemoryErrorMessage(oom, jvmMetaspaceOomNewErrorMessage);
-		} else if (isDirectOutOfMemoryError(oom)) {
-			return changeOutOfMemoryErrorMessage(oom, jvmDirectOomNewErrorMessage);
-		}
+	public static void tryEnrichOutOfMemoryError(
+			@Nullable Throwable root,
+			@Nullable String jvmMetaspaceOomNewErrorMessage,
+			@Nullable String jvmDirectOomNewErrorMessage,
+			@Nullable String jvmHeapSpaceOomNewErrorMessage) {
+		updateDetailMessage(root, t -> {
+			if (isMetaspaceOutOfMemoryError(t)) {
+				return jvmMetaspaceOomNewErrorMessage;
+			} else if (isDirectOutOfMemoryError(t)) {
+				return jvmDirectOomNewErrorMessage;
+			} else if (isHeapSpaceOutOfMemoryError(t)) {
+				return jvmHeapSpaceOomNewErrorMessage;
+			}
 
-		return oom;
+			return null;
+		});
 	}
 
 	/**
-	 * Rewrites the error message of a {@link OutOfMemoryError}.
+	 * Updates the error message of the first Throwable appearing in the cause tree of the passed root Throwable and
+	 * matching the passed Predicate by traversing the cause tree top-to-bottom.
 	 *
-	 * @param oom original {@link OutOfMemoryError}
-	 * @param newMessage new error message
-	 * @return the origianl {@link OutOfMemoryError} if it already has the new error message or
-	 * a new {@link OutOfMemoryError} with the new error message
+	 * @param root The Throwable whose cause tree shall be traversed.
+	 * @param throwableToMessage The Function based on which the new messages are generated. The function implementation
+	 *                           should return the new message. Returning <code>null</code>, in contrast, will result in
+	 *                           not updating the message for the corresponding Throwable.
 	 */
-	private static OutOfMemoryError changeOutOfMemoryErrorMessage(OutOfMemoryError oom, String newMessage) {
-		if (oom.getMessage().equals(newMessage)) {
-			return oom;
+	public static void updateDetailMessage(Throwable root, Function<Throwable, String> throwableToMessage) {
```

Comment:
if `root` can be `null`, then let's add `@Nullable`. The same for `throwableToMessage`. However, it might not be necessary to support `null` values here.